File: qbopt/model/execute.py
"""Small executable reference semantics for integer MIR.

An oracle for pass tests: run a body before and after a transform on the
same inputs and compare what each returns and stores. Anything not modelled
raises rather than guessing.

Memory is bytes keyed by segment and 16-bit offset, as the machine keys
them; `_where` derives both for every reference. A far reference's segment
is its selector's value. Otherwise the frame is SS, with `bp` zero, and a
near reference is DS -- or SS where it says it points into the frame. With
the stack in data, SS is DS. A named segment in DGROUP is DS at its base
offset; any other named cell is its own segment.
"""

from dataclasses import field
from dataclasses import dataclass
from collections.abc import Mapping
from collections.abc import Callable

from qbopt.model import mir
from qbopt.objectfile.module import Space
from iced_x86 import Register

# DS and SS. A far selector is a number, so neither is ever mistaken for one.
DS, SS = "ds", "ss"


class ExecutionError(ValueError):
    """The body uses something this executor does not model."""


@dataclass(frozen=True, slots=True)
class _Flags:
    kind: mir.Kind
    left: int
    right: int
    result: int
    width: int


@dataclass(slots=True)
class State:
    values: dict[mir.Value, int | _Flags] = field(default_factory=dict)
    memory: dict[tuple[object, int], int] = field(default_factory=dict)
    steps: int = 0
    stack: str = SS
    # A DGROUP segment's index and where it starts in DS.
    dgroup: Mapping[int, int] = field(default_factory=dict)


@dataclass(frozen=True, slots=True)
class Result:
    returned: tuple[int, ...]
    memory: dict[tuple[object, int], int]
    steps: int


type Call = Callable[[mir.Op, tuple[int, ...]], tuple[int, ...]]


def _mask(n: int, width: int) -> int:
    return n & ((1 << 8 * width) - 1)


def _signed(n: int, width: int) -> int:
    n = _mask(n, width)
    return n - (1 << 8 * width) if n >> (8 * width - 1) else n


def _compare(test: mir.Kind, left: int, right: int, width: int) -> bool:
    a, b = _signed(left, width), _signed(right, width)
    u, v = _mask(left, width), _mask(right, width)
    match test:
        case mir.Kind.EQ:
            return u == v
        case mir.Kind.NE:
            return u != v
        case mir.Kind.LT:
            return a < b
        case mir.Kind.LE:
            return a <= b
        case mir.Kind.GT:
            return a > b
        case mir.Kind.GE:
            return a >= b
        case mir.Kind.BELOW:
            return u < v
        case mir.Kind.BELOW_EQ:
            return u <= v
        case mir.Kind.ABOVE:
            return u > v
        case mir.Kind.ABOVE_EQ:
            return u >= v
    raise ExecutionError(f"no comparison {test}")


def _taken(flags: _Flags, test: mir.Kind) -> bool:
    if flags.kind is mir.Kind.SUB:
        return _compare(test, flags.left, flags.right, flags.width)
    # Logic leaves carry and overflow clear: every test is the result against zero.
    if flags.kind in (mir.Kind.AND, mir.Kind.OR, mir.Kind.XOR) or test in (mir.Kind.EQ, mir.Kind.NE):
        return _compare(test, flags.result, 0, flags.width)
    raise ExecutionError(f"{test} on the flags of {flags.kind}")


def _where(ref: mir.MemRef, state: State) -> tuple[object, int]:
    """The segment and offset of the first byte `ref` names."""
    addr = ref.addr
    if addr is None:
        raise ExecutionError(f"no address for {ref}")
    offset = addr.disp + (_integer(state, ref.base) if ref.base is not None else 0)
    if ref.segment is not None:
        return _integer(state, ref.segment), _mask(offset, 2)
    match addr.space:
        case Space.FRAME:
            segment = state.stack
        case Space.LITERAL:
            framed = ref.space is Space.FRAME or addr.segment == Register.SS
            segment = state.stack if framed else DS
        case Space.SEGMENT if addr.index in state.dgroup:
            segment, offset = DS, offset + state.dgroup[addr.index]
        case Space.SEGMENT | Space.EXTERNAL:
            segment = (addr.space, addr.index)
        case _:
            raise ExecutionError(f"no segment for {ref}")
    return segment, _mask(offset, 2)


def _load(state: State, ref: mir.MemRef, width: int) -> int:
    region, offset = _where(ref, state)
    return sum(state.memory.get((region, _mask(offset + at, 2)), 0) << 8 * at for at in range(width))


def _store(state: State, ref: mir.MemRef, n: int, width: int) -> None:
    region, offset = _where(ref, state)
    for at in range(width):
        state.memory[(region, _mask(offset + at, 2))] = n >> 8 * at & 0xFF


def _integer(state: State, value: mir.Value) -> int:
    got = state.values.get(value)
    if got is None:
        raise ExecutionError(f"{value!r} read before it is defined")
    if isinstance(got, _Flags):
        raise ExecutionError(f"{value!r} is flags, not a number")
    return got


def _read(state: State, arg: mir.Arg) -> int:
    match arg:
        case mir.Held(value, width):
            return _mask(_integer(state, value), width)
        case mir.Const(n, width):
            return _mask(n, width)
        case mir.Cell(ref):
            return _load(state, ref, ref.width)
        case mir.FrameAddress(offset, width):
            return _mask(offset, width)  # bp is zero
    raise ExecutionError(f"cannot read {arg!r}")


def _width(op: mir.Op) -> int:
    for one in (*op.results, *op.args):
        if isinstance(one, (mir.Held, mir.Const)):
            return one.width
        if isinstance(one, mir.Cell):
            return one.ref.width
    raise ExecutionError(f"no width for {op.kind}")


def _computed(op: mir.Op, args: tuple[int, ...], width: int) -> tuple[int, ...]:
    kind = op.kind
    bits = 8 * width
    match kind, args:
        case mir.Kind.COPY | mir.Kind.CONVERT | mir.Kind.ZERO_EXTEND | mir.Kind.LOAD | mir.Kind.ADDRESS, (a,):
            return (a,)
        case mir.Kind.SIGN_EXTEND, (a,):
            source = op.args[0]
            return (_signed(a, source.width) if isinstance(source, (mir.Held, mir.Const)) else a,)
        case mir.Kind.ADD, (a, b):
            return (a + b,)
        case mir.Kind.SUB, (a, b):
            return (a - b,)
        case mir.Kind.MUL, (a, b):
            return (_signed(a, width) * _signed(b, width),)
        case mir.Kind.AND, (a, b):
            return (a & b,)
        case mir.Kind.OR, (a, b):
            return (a | b,)
        case mir.Kind.XOR, (a, b):
            return (a ^ b,)
        case mir.Kind.SHL, (a, b):
            return (a << (b & 31),)
        case mir.Kind.SHR, (a, b):
            return (_mask(a, width) >> (b & 31),)
        case mir.Kind.SAR, (a, b):
            return (_signed(a, width) >> (b & 31),)
        case mir.Kind.NEG, (a,):
            return (-a,)
        case mir.Kind.NOT, (a,):
            return (~a,)
        case mir.Kind.INCREMENT, (a,):
            return (a + 1,)
        case mir.Kind.DECREMENT, (a,):
            return (a - 1,)
        case mir.Kind.DIV | mir.Kind.REM | mir.Kind.DIVMOD, (a, b):
            a, b = _signed(a, width), _signed(b, width)
            if b == 0:
                raise ExecutionError("division by zero")
            quotient = abs(a) // abs(b) * (1 if (a < 0) == (b < 0) else -1)
            remainder = a - quotient * b
            return {mir.Kind.DIV: (quotient,), mir.Kind.REM: (remainder,)}.get(kind, (quotient, remainder))
        case mir.Kind.UDIVMOD, (a, b):
            if b == 0:
                raise ExecutionError("division by zero")
            return (a // b, a % b)
        case mir.Kind.EXTRACT, (a, offset):
            return (a >> offset,)
        case mir.Kind.CONCAT, (high, low):
            low_width = op.args[1].width if isinstance(op.args[1], (mir.Held, mir.Const)) else width // 2
            return (high << 8 * low_width | _mask(low, low_width),)
    if kind in mir.MIRRORED and len(args) == 2:
        source = op.args[0]
        compared = source.width if isinstance(source, (mir.Held, mir.Const)) else width
        return (int(_compare(kind, args[0], args[1], compared)),)
    raise ExecutionError(f"{kind} with {len(args)} operands (bits {bits})")


def _executed(op: mir.Op, state: State, call: Call | None) -> None:
    if op.kind is mir.Kind.ADDRESS and len(op.args) == 1 and isinstance(op.args[0], mir.Cell):
        args = (_where(op.args[0].ref, state)[1],)
    else:
        args = tuple(_read(state, arg) for arg in op.args)
    if op.kind is mir.Kind.STORE:
        (target,) = op.results
        if not isinstance(target, mir.Cell):
            raise ExecutionError("a store without a cell")
        _store(state, target.ref, args[0], target.ref.width)
        return
    if op.kind is mir.Kind.CALL:
        if call is None:
            raise ExecutionError(f"call {op.name}")
        answers = call(op, args)
    else:
        width = _width(op)
        answers = tuple(_mask(n, width) for n in _computed(op, args, width))
    flags = [value for value in op.defines if value.flags]
    if flags and op.kind is not mir.Kind.CALL:
        width = _width(op)
        result = answers[0] if answers else _mask(args[0] - args[1], width) if op.kind is mir.Kind.SUB else args[0]
        left, right = (args + (0, 0))[:2]
        for value in flags:
            state.values[value] = _Flags(op.kind, left, right, result, width)
    for target, n in zip(op.results, answers, strict=False):
        match target:
            case mir.Held(value, width):
                state.values[value] = _mask(n, width)
            case mir.Cell(ref):
                _store(state, ref, n, ref.width)


def run(
    body: mir.MirBody,
    values: dict[mir.Value, int] | None = None,
    memory: dict[tuple[object, int], int] | None = None,
    *,
    call: Call | None = None,
    limit: int = 1_000_000,
    dgroup: Mapping[int, int] | None = None,
) -> Result:
    """Execute `body` from its entry until it returns.

    `values` are the live-in values; `memory` the bytes it starts with,
    keyed as `_where` keys them; `dgroup` each DGROUP segment's base in DS.
    """
    stack = DS if body.stack_in_data else SS
    state = State(dict(values or {}), dict(memory or {}), stack=stack, dgroup=dict(dgroup or {}))
    blocks = {block.at: block for block in body.blocks}
    for ref, constant in body.initial:
        _store(state, ref, constant.n, ref.width)
    at, came = body.entry, None
    while True:
        block = blocks[at]
        if came is not None:
            incoming = {phi.result: state.values.get(phi.incoming[came]) for phi in block.phis if came in phi.incoming}
            for result, got in incoming.items():
                if got is None:
                    raise ExecutionError(f"phi {result!r} reads an undefined value from b{came}")
                state.values[result] = got
        following = block.succ[0] if block.succ else None
        for op in block.ops:
            state.steps += 1
            if state.steps > limit:
                raise ExecutionError("step limit")
            match op.kind:
                case mir.Kind.NOTHING:
                    continue
                case mir.Kind.JUMP:
                    following = op.target if op.target is not None else following
                    break
                case mir.Kind.RETURN:
                    return Result(tuple(_read(state, arg) for arg in op.args), state.memory, state.steps)
                case mir.Kind.BRANCH:
                    (flags,) = [value for value in op.uses if value.flags]
                    got = state.values.get(flags)
                    if not isinstance(got, _Flags) or op.test is None:
                        raise ExecutionError(f"branch on {flags!r}")
                    others = [one for one in block.succ if one != op.target]
                    following = op.target if _taken(got, op.test) else (others[0] if others else op.target)
                    break
                case mir.Kind.SWITCH | mir.Kind.ESCAPE | mir.Kind.OPAQUE | mir.Kind.ARG | mir.Kind.RESULT:
                    raise ExecutionError(f"{op.kind} is not modelled")
                case _:
                    if op.floating is not None or op.kind.value.startswith("f"):
                        raise ExecutionError(f"{op.kind} is not modelled")
                    _executed(op, state, call)
        if following is None:
            raise ExecutionError(f"b{at} ends without a successor")
        at, came = following, at

File: crates/llrm-core/src/model/execute.rs
//! Small executable reference semantics for integer MIR.
//!
//! Port of `qbopt/model/execute.py`. An oracle for pass tests: run a body
//! before and after a transform on the same inputs and compare what each
//! returns and stores. Anything not modelled raises rather than guessing.
//!
//! Memory is bytes keyed by segment and 16-bit offset, as the machine keys
//! them; `_where` derives both for every reference. A far reference's segment
//! is its selector's value. Otherwise the frame is SS, with `bp` zero, and a
//! near reference is DS -- or SS where it says it points into the frame. With
//! the stack in data, SS is DS. A named segment in DGROUP is DS at its base
//! offset; any other named cell is its own segment.

// An oracle for tests; nothing in the compiler runs it.
#![cfg_attr(not(test), allow(dead_code))]

use std::collections::BTreeMap;
use std::fmt;

use iced_x86::Register;
use crate::support::hash::IndexMap;
use num_bigint::BigInt;
use num_traits::{Signed, ToPrimitive, Zero};

use crate::model::mir::{self, Arg, Kind, MemRef, MirBody, Op, Value};
use crate::objectfile::module::Space;
use crate::support::pyrepr::Repr;

/// Python's segment object: `DS`, `SS`, a far selector's number, or a named `(space, index)`.
#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub enum Segment {
    Ds,
    Ss,
    Selector(BigInt),
    Named(Space, i64),
}

/// The body uses something this executor does not model.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ExecutionError(pub String);

impl fmt::Display for ExecutionError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.0)
    }
}

impl std::error::Error for ExecutionError {}

fn error<T>(message: String) -> Result<T, ExecutionError> {
    Err(ExecutionError(message))
}

#[derive(Clone, Debug, Eq, PartialEq)]
struct _Flags {
    kind: Kind,
    left: BigInt,
    right: BigInt,
    result: BigInt,
    width: u32,
}

/// Python's `int | _Flags` value slot.
#[derive(Clone, Debug, Eq, PartialEq)]
enum Stored {
    Integer(BigInt),
    Flags(_Flags),
}

pub type Memory = IndexMap<(Segment, i64), u8>;

pub struct State {
    values: IndexMap<Value, Stored>,
    memory: Memory,
    steps: u64,
    stack: Segment,
    // A DGROUP segment's index and where it starts in DS.
    dgroup: BTreeMap<i64, i64>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Outcome {
    pub returned: Vec<BigInt>,
    pub memory: Memory,
    pub steps: u64,
}

pub type Call<'a> = dyn FnMut(&Op, &[BigInt]) -> Result<Vec<BigInt>, ExecutionError> + 'a;

fn _mask(n: &BigInt, width: u32) -> BigInt {
    n & ((BigInt::from(1_u8) << (8 * width)) - 1)
}

fn _signed(n: &BigInt, width: u32) -> BigInt {
    let n = _mask(n, width);
    if (&n >> (8 * width - 1)).is_zero() { n } else { n - (BigInt::from(1_u8) << (8 * width)) }
}

fn _compare(test: Kind, left: &BigInt, right: &BigInt, width: u32) -> Result<bool, ExecutionError> {
    let (a, b) = (_signed(left, width), _signed(right, width));
    let (u, v) = (_mask(left, width), _mask(right, width));
    Ok(match test {
        Kind::Eq => u == v,
        Kind::Ne => u != v,
        Kind::Lt => a < b,
        Kind::Le => a <= b,
        Kind::Gt => a > b,
        Kind::Ge => a >= b,
        Kind::Below => u < v,
        Kind::BelowEq => u <= v,
        Kind::Above => u > v,
        Kind::AboveEq => u >= v,
        _ => return error(format!("no comparison {test}")),
    })
}

fn _taken(flags: &_Flags, test: Kind) -> Result<bool, ExecutionError> {
    if flags.kind == Kind::Sub {
        return _compare(test, &flags.left, &flags.right, flags.width);
    }
    // Logic leaves carry and overflow clear: every test is the result against zero.
    if matches!(flags.kind, Kind::And | Kind::Or | Kind::Xor) || matches!(test, Kind::Eq | Kind::Ne) {
        return _compare(test, &flags.result, &BigInt::zero(), flags.width);
    }
    error(format!("{test} on the flags of {}", flags.kind))
}

/// The segment and offset of the first byte `ref` names.
fn _where(r#ref: &MemRef, state: &State) -> Result<(Segment, i64), ExecutionError> {
    let Some(addr) = &r#ref.addr else {
        return error(format!("no address for {}", r#ref.repr()));
    };
    let mut offset = BigInt::from(addr.disp)
        + match r#ref.base {
            Some(base) => _integer(state, base)?,
            None => BigInt::zero(),
        };
    let short = |offset: &BigInt| _mask(offset, 2).to_i64().expect("16 bits");
    if let Some(segment) = r#ref.segment {
        return Ok((Segment::Selector(_integer(state, segment)?), short(&offset)));
    }
    let segment = match addr.space {
        Space::Frame => state.stack.clone(),
        Space::Literal => {
            let framed = r#ref.space == Some(Space::Frame) || addr.segment == Register::SS;
            if framed { state.stack.clone() } else { Segment::Ds }
        }
        Space::Segment if state.dgroup.contains_key(&addr.index) => {
            offset += state.dgroup[&addr.index];
            Segment::Ds
        }
        Space::Segment | Space::External => Segment::Named(addr.space, addr.index),
        _ => return error(format!("no segment for {}", r#ref.repr())),
    };
    Ok((segment, short(&offset)))
}

fn _load(state: &State, r#ref: &MemRef, width: u32) -> Result<BigInt, ExecutionError> {
    let (region, offset) = _where(r#ref, state)?;
    Ok((0..width)
        .map(|at| {
            let byte = state.memory.get(&(region.clone(), (offset + i64::from(at)) & 0xFFFF)).copied().unwrap_or(0);
            BigInt::from(byte) << (8 * at)
        })
        .sum())
}

fn _store(state: &mut State, r#ref: &MemRef, n: &BigInt, width: u32) -> Result<(), ExecutionError> {
    let (region, offset) = _where(r#ref, state)?;
    for at in 0..width {
        let byte = ((n >> (8 * at)) & BigInt::from(0xFF_u8)).to_u8().expect("a byte");
        state.memory.insert((region.clone(), (offset + i64::from(at)) & 0xFFFF), byte);
    }
    Ok(())
}

fn _integer(state: &State, value: Value) -> Result<BigInt, ExecutionError> {
    match state.values.get(&value) {
        None => error(format!("{} read before it is defined", value.repr())),
        Some(Stored::Flags(_)) => error(format!("{} is flags, not a number", value.repr())),
        Some(Stored::Integer(got)) => Ok(got.clone()),
    }
}

fn _read(state: &State, arg: &Arg) -> Result<BigInt, ExecutionError> {
    match arg {
        Arg::Held(held) => Ok(_mask(&_integer(state, held.value)?, held.width)),
        Arg::Const(constant) => Ok(_mask(&constant.n, constant.width)),
        Arg::Cell(cell) => _load(state, &cell.r#ref, cell.r#ref.width),
        Arg::FrameAddress(frame) => Ok(_mask(&BigInt::from(frame.offset), frame.width)), // bp is zero
        _ => error(format!("cannot read {}", arg.repr())),
    }
}

fn _width(op: &Op) -> Result<u32, ExecutionError> {
    for one in op.results.iter().chain(&op.args) {
        match one {
            Arg::Held(held) => return Ok(held.width),
            Arg::Const(constant) => return Ok(constant.width),
            Arg::Cell(cell) => return Ok(cell.r#ref.width),
            _ => {}
        }
    }
    error(format!("no width for {}", op.kind))
}

/// Python's `.width` of a `Held | Const` operand.
fn _operand_width(arg: &Arg) -> Option<u32> {
    match arg {
        Arg::Held(held) => Some(held.width),
        Arg::Const(constant) => Some(constant.width),
        _ => None,
    }
}

fn _computed(op: &Op, args: &[BigInt], width: u32) -> Result<Vec<BigInt>, ExecutionError> {
    let kind = op.kind;
    let bits = 8 * width;
    let shift = |b: &BigInt| (b & BigInt::from(31_u8)).to_usize().expect("five bits");
    match (kind, args) {
        (Kind::Copy | Kind::Convert | Kind::ZeroExtend | Kind::Load | Kind::Address, [a]) => {
            return Ok(vec![a.clone()]);
        }
        (Kind::SignExtend, [a]) => {
            return Ok(vec![_operand_width(&op.args[0]).map_or_else(|| a.clone(), |source| _signed(a, source))]);
        }
        (Kind::Add, [a, b]) => return Ok(vec![a + b]),
        (Kind::Sub, [a, b]) => return Ok(vec![a - b]),
        (Kind::Mul, [a, b]) => return Ok(vec![_signed(a, width) * _signed(b, width)]),
        (Kind::And, [a, b]) => return Ok(vec![a & b]),
        (Kind::Or, [a, b]) => return Ok(vec![a | b]),
        (Kind::Xor, [a, b]) => return Ok(vec![a ^ b]),
        (Kind::Shl, [a, b]) => return Ok(vec![a << shift(b)]),
        (Kind::Shr, [a, b]) => return Ok(vec![_mask(a, width) >> shift(b)]),
        (Kind::Sar, [a, b]) => return Ok(vec![_signed(a, width) >> shift(b)]),
        (Kind::Neg, [a]) => return Ok(vec![-a]),
        (Kind::Not, [a]) => return Ok(vec![!a]),
        (Kind::Increment, [a]) => return Ok(vec![a + 1]),
        (Kind::Decrement, [a]) => return Ok(vec![a - 1]),
        (Kind::Div | Kind::Rem | Kind::Divmod, [a, b]) => {
            let (a, b) = (_signed(a, width), _signed(b, width));
            if b.is_zero() {
                return error("division by zero".to_owned());
            }
            let sign = if a.is_negative() == b.is_negative() { 1 } else { -1 };
            let quotient = a.abs() / b.abs() * sign;
            let remainder = &a - &quotient * &b;
            return Ok(match kind {
                Kind::Div => vec![quotient],
                Kind::Rem => vec![remainder],
                _ => vec![quotient, remainder],
            });
        }
        (Kind::Udivmod, [a, b]) => {
            if b.is_zero() {
                return error("division by zero".to_owned());
            }
            return Ok(vec![a / b, a % b]);
        }
        (Kind::Extract, [a, offset]) => return Ok(vec![a >> offset.to_usize().expect("a bit offset")]),
        (Kind::Concat, [high, low]) => {
            let low_width = _operand_width(&op.args[1]).unwrap_or(width / 2);
            return Ok(vec![(high << (8 * low_width)) | _mask(low, low_width)]);
        }
        _ => {}
    }
    if mir::MIRRORED(kind).is_some() && args.len() == 2 {
        let compared = _operand_width(&op.args[0]).unwrap_or(width);
        return Ok(vec![BigInt::from(u8::from(_compare(kind, &args[0], &args[1], compared)?))]);
    }
    error(format!("{kind} with {} operands (bits {bits})", args.len()))
}

fn _executed(op: &Op, state: &mut State, call: Option<&mut Call<'_>>) -> Result<(), ExecutionError> {
    let args = match (op.kind, op.args.as_slice()) {
        (Kind::Address, [Arg::Cell(cell)]) => vec![BigInt::from(_where(&cell.r#ref, state)?.1)],
        _ => op.args.iter().map(|arg| _read(state, arg)).collect::<Result<Vec<_>, _>>()?,
    };
    if op.kind == Kind::Store {
        let [target] = op.results.as_slice() else {
            panic!("ValueError: a store has one result");
        };
        let Arg::Cell(target) = target else {
            return error("a store without a cell".to_owned());
        };
        return _store(state, &target.r#ref, &args[0], target.r#ref.width);
    }
    let answers = if op.kind == Kind::Call {
        let Some(call) = call else {
            return error(format!("call {}", op.name));
        };
        call(op, &args)?
    } else {
        let width = _width(op)?;
        _computed(op, &args, width)?.iter().map(|n| _mask(n, width)).collect()
    };
    let flags = op.defines.iter().filter(|value| value.flags).copied().collect::<Vec<_>>();
    if !flags.is_empty() && op.kind != Kind::Call {
        let width = _width(op)?;
        let result = match answers.first() {
            Some(first) => first.clone(),
            None if op.kind == Kind::Sub => _mask(&(&args[0] - &args[1]), width),
            None => args[0].clone(),
        };
        let padded = args.iter().cloned().chain([BigInt::zero(), BigInt::zero()]).collect::<Vec<_>>();
        let (left, right) = (padded[0].clone(), padded[1].clone());
        for value in flags {
            let stored = _Flags { kind: op.kind, left: left.clone(), right: right.clone(), result: result.clone(), width };
            state.values.insert(value, Stored::Flags(stored));
        }
    }
    for (target, n) in op.results.iter().zip(&answers) {
        match target {
            Arg::Held(held) => {
                state.values.insert(held.value, Stored::Integer(_mask(n, held.width)));
            }
            Arg::Cell(cell) => _store(state, &cell.r#ref, n, cell.r#ref.width)?,
            _ => {}
        }
    }
    Ok(())
}

/// Execute `body` from its entry until it returns.
///
/// `values` are the live-in values; `memory` the bytes it starts with,
/// keyed as `_where` keys them; `dgroup` each DGROUP segment's base in DS.
pub fn run(
    body: &MirBody,
    values: &IndexMap<Value, BigInt>,
    memory: &Memory,
    mut call: Option<&mut Call<'_>>,
    limit: u64,
    dgroup: &BTreeMap<i64, i64>,
) -> Result<Outcome, ExecutionError> {
    let stack = if body.stack_in_data { Segment::Ds } else { Segment::Ss };
    let mut state = State {
        values: values.iter().map(|(value, n)| (*value, Stored::Integer(n.clone()))).collect(),
        memory: memory.clone(),
        steps: 0,
        stack,
        dgroup: dgroup.clone(),
    };
    let blocks = body.blocks.iter().map(|block| (block.at, block)).collect::<BTreeMap<_, _>>();
    for (r#ref, constant) in &body.initial {
        _store(&mut state, r#ref, &constant.n, r#ref.width)?;
    }
    let (mut at, mut came) = (body.entry, None::<i64>);
    loop {
        let block = blocks[&at];
        if let Some(came) = came {
            let incoming = block
                .phis
                .iter()
                .filter_map(|phi| {
                    let source = phi.incoming.get(&came)?;
                    Some((phi.result, state.values.get(source).cloned()))
                })
                .collect::<IndexMap<_, _>>();
            for (result, got) in incoming {
                let Some(got) = got else {
                    return error(format!("phi {} reads an undefined value from b{came}", result.repr()));
                };
                state.values.insert(result, got);
            }
        }
        let mut following = block.succ.first().copied();
        for op in &block.ops {
            state.steps += 1;
            if state.steps > limit {
                return error("step limit".to_owned());
            }
            match op.kind {
                Kind::Nothing => continue,
                Kind::Jump => {
                    following = op.target.or(following);
                    break;
                }
                Kind::Return => {
                    let returned = op.args.iter().map(|arg| _read(&state, arg)).collect::<Result<Vec<_>, _>>()?;
                    return Ok(Outcome { returned, memory: state.memory, steps: state.steps });
                }
                Kind::Branch => {
                    let [flags] = op.uses.iter().filter(|value| value.flags).collect::<Vec<_>>()[..] else {
                        panic!("ValueError: a branch reads one flags value");
                    };
                    let got = match state.values.get(flags) {
                        Some(Stored::Flags(got)) if op.test.is_some() => got,
                        _ => return error(format!("branch on {}", flags.repr())),
                    };
                    let test = op.test.expect("checked above");
                    let others = block.succ.iter().filter(|one| Some(**one) != op.target).collect::<Vec<_>>();
                    following = if _taken(got, test)? { op.target } else { others.first().map(|one| **one).or(op.target) };
                    break;
                }
                Kind::Switch | Kind::Escape | Kind::Opaque | Kind::Arg | Kind::Result => {
                    return error(format!("{} is not modelled", op.kind));
                }
                _ => {
                    if op.floating.is_some() || op.kind.as_str().starts_with('f') {
                        return error(format!("{} is not modelled", op.kind));
                    }
                    _executed(op, &mut state, call.as_deref_mut())?;
                }
            }
        }
        let Some(next) = following else {
            return error(format!("b{at} ends without a successor"));
        };
        (at, came) = (next, Some(at));
    }
}

#[cfg(test)]
#[path = "execute_tests.rs"]
mod tests;
